### ISSUE TITLE
ModelResolver: How to ref enum type with enum as schema component?

### ISSUE THREAD

**Issue by mtraynham** [swagger-api/swagger-core#3124]
In the ModelResolver, there is this [piece of code](https://github.com/swagger-api/swagger-core/blob/master/modules/swagger-core/src/main/java/io/swagger/v3/core/jackson/ModelResolver.java#L241-L244).

Given an enum property and model class 
```java
public class Test {
    public State state;
    public enum State { UP, DOWN }
}
```

This generates a `string` property with the enum values exploded out.
```json
      "Test" : {
        "type" : "object",
        "properties" : {
          "state" : {
            "type" : "string",
            "enum" : [ "UP", "DOWN" ]
          }
        }
      }
```
---
According to the Enums [documentation](https://swagger.io/docs/specification/data-models/enums/), under the section **Reusable Enums**, it could instead be formatted like so:

```json
      "Test" : {
        "type" : "object",
        "properties" : {
          "state" : {
            "$ref": "#/components/schemas/State"
          }
        }
      },
      "State" : {
        "type" : "string",
        "enum" : [ "UP", "DOWN" ]
      }
```

How does one achieve this?

---
I tried the following variants:

### 1
```java
public class Test {
    public State state;
    
    @Schema
    public enum State { UP, DOWN }
}
```
```json
      "Test" : {
        "type" : "object",
        "properties" : {
          "state" : {
            "type" : "string",
            "enum" : [ "UP", "DOWN" ]
          }
        }
      }
```

### 2
```java
public class Test {
    public State state;

    @Schema(
        type = "string",
        implementation = String.class,
        allowableValues = ["UP", "DOWN"]
    )
    public enum State { UP, DOWN }
}

```
```json
      "Test" : {
        "type" : "object",
        "properties" : {
          "state" : {
            "type" : "string",
            "enum" : [ "UP", "DOWN" ]
          }
        }
      }
```

### 3
```java
public class Test {
    public State state;

    @Schema(
        type = "string",
        implementation = String.class,
        allowableValues = ["UP", "DOWN"],
        ref = "#/components/schemas/State"
    )
    public enum State { UP, DOWN }
}
```
```json
      "Test" : {
        "type" : "object",
        "properties" : {
          "state" : {
            "$ref" : "#/components/schemas/State"
          }
        }
      },
      "State" : {
        "$ref" : "#/components/schemas/State"
      }
```

### 4
```java
public class Test {
    @Schema(ref = "#/components/schemas/State")
    public State state;

    @Schema(
        type = "string",
        implementation = String.class,
        allowableValues = ["UP", "DOWN"]
    )
    public enum State { UP, DOWN }
}
```
```json
      "Test" : {
        "type" : "object",
        "properties" : {
          "state" : {
            "type" : "string"
          }
        }
      }
```

### 5
```java
public class Test {
    @Schema(ref = "#/components/schemas/State")
    public State state;

    @Schema(
        type = "string",
        allowableValues = ["UP", "DOWN"]
    )
    public enum State { UP, DOWN }
}
```
```json
      "Test" : {
        "type" : "object",
        "properties" : {
          "state" : {
            "type" : "string",
            "enum" : [ "UP", "DOWN" ]
          }
        }
      }
```

### 6
```java
public class Test {
    @Schema(ref = "#/components/schemas/State")
    public State state;

    @Schema(
        implementation = String.class,
        allowableValues = ["UP", "DOWN"]
    )
    public enum State { UP, DOWN }
}
```
```json
      "Test" : {
        "type" : "object",
        "properties" : {
          "state" : {
            "type" : "string"
          }
        }
      }
```

**Comment by jvs64893** [comment]
I just ran into the same problem. This is especially bad for classes that expose access rights, e.g. read / readwrite. Then you get X repetitions of the same enum, which is rather cumbersome to use.

**Comment by xiaoxiangmoe** [comment]
@frantuma Is there any progress in this feature? 😊

**Comment by mtraynham** [comment]
I was taking a look through the open PRs, I think #2996 handles this (original issue reported with #2994).

Changes are performed on the following lines:
https://github.com/swagger-api/swagger-core/pull/2996/files#diff-d4e9202b05579faab0e8109441f491cfR243
https://github.com/swagger-api/swagger-core/pull/2996/files#diff-098e4372496f1a4de0a865a99aafdad8R75

**Comment by frantuma** [comment]
Thanks for reporting and analyzing, and for the PR; this is fixed in #3352 which includes and replaces #3294

**Comment by devdavidkarlsson** [comment]
```
public class Test {
    public State state;

    @Schema(enumAsRef=true)
    public enum State { UP, DOWN }
}
```
